book/src/overview.md

```diff
@@ -147,12 +150,13 @@ Tracked functions can return any clone-able type. A clone is required since, whe
 
 **Tracked structs** are intermediate structs created during your computation.
 Like inputs, their fields are stored inside the database, and the struct itself just wraps an id.
-Unlike inputs, they can only be created inside a tracked function, and their fields can never change once they are created.
-Getter methods are provided to read the fields, but there are no setter methods[^specify]. Example:
+Unlike inputs, they can only be created inside a tracked function, and their fields can never change once they are created (until the next revision, at least).
+Getter methods are provided to read the fields, but there are no setter methods. 
+Example:
 
 ```rust
 #[salsa::tracked]
-struct Ast {
+struct Ast<'db> {
```

Comment:
Does the addition of the `db` lifetime also allow queries to return data that reference the DB?

One use case that we have is that we need a mapping from `AstNode` -> `Id` where `Id` for example uniquely identifie's a scope, or a symbol in the program. 

The challenge we're facing is that our `Ast` doesn't use `Arc`s internally, thus cloning a `Node` always clones the entire sub-tree. Our "work-around" for this is to keep hold to the AST's root (wrapped in an Arc) and store a raw pointer referencing the actual node. This works pretty well, but requires heavy use of `Arc`s (a lot of cloning). I "think" your changes would allow us to directly store a `&'db Expr` instead. 

If not, then the "work-around" would just be to make the `AstNode` -> `Id` map a salsa tracked so that we get access to the db lifetime

Reply:
I did intend to support storing `&T` references like that, but it's a subtle case, and I've gone back and forth on whether it works with the stacked borrows rules etc. 

Suppose you do `let f = ast.field(db)` in R1 and it yields a `&'db Foo` (reference to some field of `ast`) and then you store that in the database as the result of a query (or part of the result). Now say we start a new revision R2 and, in R2, the value of `field` changes. This means that `f` (considered as a pointer) still points to the same memory, but the value behind `f` has changed. There are two challenges: (a) under the stacked borrow rules, it is UB to use `f` again; (b) should we consider functions that were dependent on `f` as needing to be re-executed?

I've tried to write an exploration of this question in this comment like 3 times but it keeps getting unwieldy. I think I will defer it for documentation or in-person discussion, it's a good one. I'm not entirely sure if and under what conditions this can be made to be safe at the moment. =)

That said, I also want to mention a feature I've been considering that I think is may help with your use case. The idea would be to make it easy to have a value that carries a memory arena and references into that arena. This is meant to model things like MIR, where we have some data structure that represents a function, and to allow it to go through phases where it is changed and updated, but without requiring everything to be in vectors nor requiring everything to be cloned constantly. I'm not sure the ergonomics exactly but the idea is roughly that you can declare a struct with two lifetimes...

```rust
#[in_arena(AstRoot)]
struct Ast<'ast, 'db: 'ast> {
   data: AstData<'ast, 'db>,
   children: Vec<&'ast Ast<'ast, 'db>>,
}
```

...and the procedural macro will create a type `AstRoot` that "hides" the first one:

```rust
struct AstRoot<'db> {
    arena: Arc<MemoryArena>,
    root: &'static Ast<'static 'static>, // <-- the lifetimes here are obviously lies
}
```

Later you can do `root.open(|r| { .. })` to work with the data. One of the goals is that you can create new, derived values based on the same arena that have different pointers -- so e.g. it should be possible to extra subvalues from the tree. Each of them would carry a reference count to the same base arena.

Reply:
> That said, I also want to mention a feature I've been considering that I think is may help with your use case. The idea would be to make it easy to have a value that carries a memory arena and references into that arena. 

Yeah, that sounds very similar to our "work-around", except that it is more flexible and the unsafety is handled by salsa instead of sprinkled through our code.